src/vm/functions/principals.rs

```diff
@@ -38,13 +66,199 @@ pub fn special_is_standard(
+            .expect("FAIL: Failed to initialize tuple."),
+        )),
+    })
+}
+
+pub fn special_principal_parse(
+    args: &[SymbolicExpression],
+    env: &mut Environment,
+    context: &LocalContext,
+) -> Result<Value> {
+    check_argument_count(1, args)?;
+    runtime_cost(ClarityCostFunction::Unimplemented, env, 0)?;
+    let principal = eval(&args[0], env, context)?;
+
+    let (version_byte, hash_bytes) = match principal {
+        Value::Principal(PrincipalData::Standard(StandardPrincipalData(version, bytes))) => {
+            (version, bytes)
+        }
+        Value::Principal(PrincipalData::Contract(QualifiedContractIdentifier { issuer, name })) => {
+            (issuer.0, issuer.1)
+        }
+        _ => {
+            return Err(CheckErrors::TypeValueError(TypeSignature::PrincipalType, principal).into())
+        }
+    };
+
+    // `version_byte_is_valid` determines whether the returned `Response` is through the success
+    // channel or the error channel.
+    let version_byte_is_valid = version_matches_current_network(version_byte, env.global_context);
+
+    let tuple = create_principal_parse_tuple(version_byte, &hash_bytes);
+    Ok(Value::Response(ResponseData {
+        committed: version_byte_is_valid,
+        data: Box::new(tuple),
+    }))
+}
+
+pub fn special_principal_construct(
+    args: &[SymbolicExpression],
+    env: &mut Environment,
+    context: &LocalContext,
+) -> Result<Value> {
+    check_argument_count(2, args)?;
+    runtime_cost(ClarityCostFunction::Unimplemented, env, 0)?;
+    let version = eval(&args[0], env, context)?;
+    let hash_bytes = eval(&args[1], env, context)?;
+
+    // Check the version byte.
+    let verified_version = match version {
+        Value::Sequence(SequenceData::Buffer(BuffData { ref data })) => data,
+        _ => {
+            return {
+                // This is an aborting error because this should have been caught in analysis pass.
+                Err(CheckErrors::TypeValueError(
+                    TypeSignature::SequenceType(SequenceSubtype::BufferType(BufferLength(1))),
+                    version,
+                )
+                .into())
+            };
+        }
+    };
+
+    // This is an aborting error because this should have been caught in analysis pass.
+    if verified_version.len() > 1 {
+        return Err(CheckErrors::TypeValueError(
+            TypeSignature::SequenceType(SequenceSubtype::BufferType(BufferLength(1))),
+            version,
+        )
+        .into());
+    }
+
+    // If the version byte buffer has 0 bytes, this is a recoverable error, because it wasn't the
+    // job of the type system.
+    if verified_version.len() < 1 {
+        // do some kind of error
+        return Ok(create_principal_true_error_response(1));
+    }
+
+    // Assume: verified_version.len() == 1
+    let version_byte = (*verified_version)[0];
+
+    // If the version byte is >= 32, this is a recoverable error, because it wasn't the job of the
```

Comment:
Why 32?

Reply:
comes from this I thought https://github.com/blockstack/stacks-blockchain/blob/master/src/address/c32.rs

can the byte really be anything?

Reply:
Yes